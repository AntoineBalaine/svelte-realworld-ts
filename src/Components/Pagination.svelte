<script lang="ts">
	import { page } from "$app/stores";

	let pages: number;
	let pageNumber: number;
	page.subscribe((thing) => {
		pageNumber = thing.data.pageNumber;
		pages = thing.data.pages;
	});

	const findPaginationRange = (pageNumber: number, pagesCount: number): number[] => {
		const left = pageNumber - 1 < 4 ? pageNumber - 1 : 4;
		const right = pagesCount - pageNumber < 4 ? pagesCount - pageNumber : 4;
		const leftIdx = pageNumber - left;
		const rightIdx = pageNumber + right;
		const range = Array<number>();
		for (let i = leftIdx; i <= rightIdx; i++) {
			range.push(i);
		}
		return range;
	};

	let paginationRange = Array<number>();
	$: {
		paginationRange = findPaginationRange(pageNumber, pages);
	}
</script>

<nav aria-label="Page navigation example">
	<ul class="pagination" data-testid="pagination-list">
		{#if pageNumber - 1 > 0}
			<li class="page-item">
				<a class="page-link" href="/?page={pageNumber - 1}">Previous</a>
			</li>
		{/if}
		{#each paginationRange as paginationElement, i (paginationElement)}
			<li class="page-item" class:active={pageNumber === paginationElement}>
				<a class="page-link" href="/?page={paginationElement}"
					>{i < paginationRange.length - 1 ? paginationElement : "â€¦"}</a
				>
			</li>
		{/each}

		{#if pageNumber < pages}
			<li class="page-item"><a class="page-link" href="/?page={pageNumber + 1}">Next</a></li>
		{/if}
	</ul>
</nav>
