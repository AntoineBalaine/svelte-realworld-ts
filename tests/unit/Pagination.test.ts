import { render } from "@testing-library/svelte";
import { describe, test, expect } from "vitest";
import Pagination from "src/components/Pagination.svelte";

import type { ComponentProps } from "svelte";

describe("Pagination Component", () => {
	test("should render the component", () => {
		const props: ComponentProps<Pagination> = {};
		const paginationPreview = render(Pagination, props);

		expect(paginationPreview.getByTestId("pagination-list")).toBeDefined();

		expect(paginationPreview.container).toMatchSnapshot();
	});
});
