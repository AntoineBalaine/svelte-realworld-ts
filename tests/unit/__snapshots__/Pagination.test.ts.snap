// Vitest Snapshot v1

exports[`Pagination Component > should render the component 1`] = `
<body>
  <div>
    <nav
      aria-label="Page navigation example"
    >
      <ul
        class="pagination"
        data-testid="pagination-list"
      >
         
         
      </ul>
    </nav>
    <!--&lt;Pagination&gt;-->
  </div>
</body>
`;
